aesara/tensor/basic.py

```diff
@@ -1817,19 +1818,148 @@ def extract_constant(x, elemwise=True, only_process_constants=False):
     return x
 
 
-def transpose(x, axes=None):
-    """
-    Reorder the dimensions of x. (Default: reverse them)
+class TransposeOp(Op):
+    def make_node(self, input_arr, new_order=None):
+        input_arr = as_tensor_variable(input_arr)
+        ib = list(input_arr.type.broadcastable)
 
-    This is a macro around dimshuffle that matches the numpy.transpose function.
+        if new_order is None:
+            new_order = as_tensor_variable(np.flip(np.arange(input_arr.ndim)))
+            ib.reverse()
+            output = TensorType(dtype=input_arr.type.dtype, broadcastable=tuple(ib))()
+            if input_arr.name:
+                output.name = input_arr.name + ".T"
+        else:
+            from aesara.tensor.var import TensorConstant
 
-    """
-    if axes is None:
-        axes = list(range((x.ndim - 1), -1, -1))
-    ret = DimShuffle(x.broadcastable, axes)(x)
-    if x.name and axes == list(range((x.ndim - 1), -1, -1)):
-        ret.name = x.name + ".T"
-    return ret
+            assert isinstance(new_order, TensorConstant)
+            # TODO: Try to remove this eval
+            _new_order = list(new_order.eval())
+            ob = []
+            if len(_new_order) != len(ib):
+                raise ValueError(
+                    f"Expected {len(ib)} number of values, found {len(_new_order)}."
+                )
+            for a in _new_order:
+                if a >= len(ib):
+                    raise ValueError("Found value greater than number of axes.")
+                if a < 0:
+                    raise ValueError(
+                        "Negative reference to axis, (e.g. axis = -1) is not allowed."
+                    )
+                axis_is_bdcast = True if a != 1 else False
+                ob.append(axis_is_bdcast)
+            output = TensorType(dtype=input_arr.type.dtype, broadcastable=ob)()
+
+        return Apply(self, [input_arr, new_order], [output])
+
+    def perform(self, node, inputs_storage, output_storage):
+        (input_arr, new_order) = inputs_storage
+        (storage,) = output_storage
+        res = input_arr.transpose(new_order)
+        storage[0] = np.asarray(res)
+
+    def grad(self, inp, grads):
```

Comment:
This grad implementation (a stripped down version of `DimShuffle`s grad implementation) seems to be going wrong somewhere. Can someone explain how gradients operate w.r.t. to such `Ops` and how are they verified ?

Reply:
Sorry, I seem to have missed this one, but the easiest way to verify an `Op.[grad, L_op, R_op]` implementation is with `tests.unittest_tools.verify_grad`.  See [here](https://aesara.readthedocs.io/en/latest/extending/creating_an_op.html#testing-the-gradient) for more testing recommendations (and, please, feel free to update it if you see something missing/wrong).